rfcs/1111-first-class-virtual-device.md

```diff
@@ -0,0 +1,146 @@
+
+Additionally, passes that move or introduce expressions will need to cooperate with device annotation (i.e., figure out the correct device themselves and insert it in any expressions they create), or device planning will need to be run again after the pass (similar to the type inference pass). In the future, we'd like to introduce a 'lite' version of the device planning pass that "fills in the gaps" left by these passes. It will propagate virtual device information strictly upwards and will have much less overhead than the full device planning pass.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The current implementation is the main alternative to making virtual devices first class. We actually considered making virtual devices first class while designing the current implementation, but we decided against it because of challenges propagating the virtual devices correctly, and we weren't sure how fundamental the virtual devices were.
+
+Another option we considered was adding the virtual device information as an attribute to each node instead of adding it as a first-class field on each node. However, not all Relay nodes have attributes, so we would have to add attributes for all the Relay nodes as well. This is also an invasive change, and we'd have to make sure attributes are propagated everywhere correctly. So this option also comes with similar challenges. And, since we'll need virtual device on every node anyways, it makes sense to just add a field for it.
+
+# Prior art
+[prior-art]: #prior-art
+
+Pytorch tensor expressions have a device field stored directly on them.
+
+Additionally, checked_type_ is a first class field on Relay nodes, so there is precedent for storing information directly on nodes.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+The name of the virtual device is up for discussion. We've considered renaming SEScope to VirtualDevice, and it seems like that is the name we will eventually go with. If we go with a different name for SEScope, the name of virtual_device_ may be different.
+
+I also need to confirm that it is feasible to propagate all spans and virtual devices through most passes. I will do this by picking a pass and making sure I can get it "span clean", meaning that it preserves all spans and propagates them correctly.
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+I'll also need to change the machinery of the device planner after this field is introduced. The basic idea is that we won't have any on_device ops after device planning anymore, all the virtual device information will be stored directly in the virtual_device_ field. I won't go into much detail about this here for the sake of limiting the scope of this discussion. Also, we will probably introduce a 'lite' version of device planning that can "fill in the gaps" in device annotations that may be left by some passes.
```

Comment:
It's also worthwhile to think about how to leverage this mechanism to simplify the BYOC flow. Currently we use `compiler_begin` and `compiler_end` to annotate the graph and face two major challenges:
1. The graph size will grow to at most 3x in the case that every op is annotated as an offloadable subgraph.
2. It's non-trivial to manipulate the compiler regions.

Ideally if we could simply annotate each call node with `virtual_device_`, we don't need `compiler_begin` and `compiler_end` anymore, but at this moment it's not clear to me about how to represent a region (subgraph) of ops on the same device. Maybe a compromise approach is we leverage `virtual_device_` in AnnotateTarget, and insert `compiler_begin/end` accordingly in MergeCompilerRegion, but this topic seems out of scope of this RFC.

Reply:
@mbs-octoml Could you take this one? Thanks!

Reply:
That was my thinking also Cody, and I started to lay it out in https://github.com/apache/tvm-rfcs/pull/38. That's somewhat moribund while we get more of the leg work in place (this change included), but would like to come back to it in Q1 with a focus on making BYOC more 'first class' (as well as address the duplication and efficiency concerns you mention.) Will be sure to ping you when that comes alive again, would very much welcome your help.

Reply:
Thanks for answering. Yeah I'd be happy to help on this at the time.